src/arch/cuda/hybridAPI/hapi_impl.cpp

```diff
@@ -519,6 +627,171 @@ hapiWorkRequest::hapiWorkRequest() :
+      CmiError("Failure at ftruncate");
+      goto shm_cleanup;
+    }
+  }
+
+  // Busywait until file is properly sized
+  do {
+    if (fstat(CsvAccess(gpu_manager).shm_file, &shm_file_stat) != 0) {
+      CmiError("Failure at fstat");
+      goto shm_cleanup;
+    }
+  } while (shm_file_stat.st_size != CsvAccess(gpu_manager).shm_size);
+
+  // Load into memory
+  CsvAccess(gpu_manager).shm_ptr = mmap(0, CsvAccess(gpu_manager).shm_size,
+      PROT_READ | PROT_WRITE, MAP_SHARED, CsvAccess(gpu_manager).shm_file, 0);
+  if (CsvAccess(gpu_manager).shm_ptr == (void*)-1) {
+    CmiError("Failure at mmap");
+    goto shm_cleanup;
+  }
+
+  // Store pointer to my process' portion of the shared memory region
+  CsvAccess(gpu_manager).shm_my_ptr = (void*)((char*)CsvAccess(gpu_manager).shm_ptr +
+      CsvAccess(gpu_manager).shm_chunk_size * CsvAccess(gpu_manager).device_count *
+      CmiMyNodeRankLocal());
+
+  // Allocate memory for local storage
+  for (int i = 0; i < CsvAccess(gpu_manager).device_count_on_physical_node; i++) {
+    CsvAccess(gpu_manager).cuda_ipc_device_infos.emplace_back();
+  }
+
+  return;
+
+shm_cleanup:
+  shmCleanup();
+  CmiAbort("Failure in shared memory region creation");
+}
+
+// Clean up shared memory region
+// Invoked by PE rank 0 of each process
+void shmCleanup() {
+  if (CsvAccess(gpu_manager).shm_ptr != NULL) {
+    munmap(CsvAccess(gpu_manager).shm_ptr, CsvAccess(gpu_manager).shm_size);
+  }
+
+  if (CsvAccess(gpu_manager).shm_file != -1) {
+    close(CsvAccess(gpu_manager).shm_file);
+  }
+
+  if (CsvAccess(gpu_manager).shm_name != NULL) {
+    shm_unlink(CsvAccess(gpu_manager).shm_name);
+    delete CsvAccess(gpu_manager).shm_name;
+    CsvAccess(gpu_manager).shm_name = NULL;
+  }
+}
+
+// Create CUDA IPC handles and populate shared memory region
+// Invoked by all PEs
+void ipcHandleCreate() {
+  // Only device reps should continue to perform the following operations
+  // so that they are done only once per device
+  if (!CpvAccess(device_rep)) return;
+
+  // Create CUDA IPC memory handle
+  CmiAssert(CsvAccess(gpu_manager).device_managers[CpvAccess(my_device)].comm_buffer);
+  cudaIpcMemHandle_t* shm_mem_handle = (cudaIpcMemHandle_t*)((char*)CsvAccess(gpu_manager).shm_my_ptr +
+      CsvAccess(gpu_manager).shm_chunk_size * CpvAccess(my_device));
+  void* device_ptr = (void*)(CsvAccess(gpu_manager).device_managers[CpvAccess(my_device)].comm_buffer->base_ptr);
+  hapiCheck(cudaIpcGetMemHandle(shm_mem_handle, device_ptr));
+
+  // Create CUDA IPC events and corresponding handles
+  cuda_ipc_event_shared* shm_event_shared = (cuda_ipc_event_shared*)((char*)shm_mem_handle + sizeof(cudaIpcMemHandle_t));
+  int device_index = CsvAccess(gpu_manager).device_count * CmiMyNodeRankLocal() + CpvAccess(my_device);
+  cuda_ipc_device_info& my_device_info = CsvAccess(gpu_manager).cuda_ipc_device_infos[device_index];
+
+  my_device_info.event_pool_flags = new int[CsvAccess(gpu_manager).ipc_event_pool_size];
```

Comment:
Is there a good reason for all of gpu_manager's arrays to be bare pointers rather than std::string and std::vector?

Reply:
Not really, at least the ones in this patch are mostly artifacts from referring to MVAPICH's C-based code.
Will go through them and C++-fy them if I can.